transmart-core-db/src/main/groovy/org/transmartproject/db/clinical/ConstraintAccessChecker.groovy

```diff
@@ -0,0 +1,186 @@
+package org.transmartproject.db.clinical
+
+import groovy.transform.CompileDynamic
+import groovy.transform.CompileStatic
+import org.transmartproject.core.exceptions.AccessDeniedException
+import org.transmartproject.core.multidimquery.query.BiomarkerConstraint
+import org.transmartproject.core.multidimquery.query.Combination
+import org.transmartproject.core.multidimquery.query.ConceptConstraint
+import org.transmartproject.core.multidimquery.query.ConstraintBuilder
+import org.transmartproject.core.multidimquery.query.FieldConstraint
+import org.transmartproject.core.multidimquery.query.ModifierConstraint
+import org.transmartproject.core.multidimquery.query.MultipleSubSelectionsConstraint
+import org.transmartproject.core.multidimquery.query.Negation
+import org.transmartproject.core.multidimquery.query.NullConstraint
+import org.transmartproject.core.multidimquery.query.PatientSetConstraint
+import org.transmartproject.core.multidimquery.query.RelationConstraint
+import org.transmartproject.core.multidimquery.query.RowValueConstraint
+import org.transmartproject.core.multidimquery.query.StudyNameConstraint
+import org.transmartproject.core.multidimquery.query.StudyObjectConstraint
+import org.transmartproject.core.multidimquery.query.SubSelectionConstraint
+import org.transmartproject.core.multidimquery.query.TemporalConstraint
+import org.transmartproject.core.multidimquery.query.TimeConstraint
+import org.transmartproject.core.multidimquery.query.TrueConstraint
+import org.transmartproject.core.multidimquery.query.ValueConstraint
+import org.transmartproject.core.ontology.MDStudy
+import org.transmartproject.core.querytool.QueryResult
+import org.transmartproject.core.users.AuthorisationChecks
+import org.transmartproject.core.users.LegacyAuthorisationChecks
+import org.transmartproject.core.users.PatientDataAccessLevel
+import org.transmartproject.core.users.User
+import org.transmartproject.db.i2b2data.Study
+import org.transmartproject.db.multidimquery.query.InvalidQueryException
+import org.transmartproject.db.querytool.QtQueryResultInstance
+
+/**
+ * Checks if the user has sufficient permissions to execute a query for a constraint.
+ * For constraints that refer to studies, concepts and patient sets, it is checked
+ * if the user has sufficient level of access to the study, to a study associated with
+ * the concept, or is the owner of the patient set, respectively.
+ */
+@CompileStatic
+class ConstraintAccessChecker extends ConstraintBuilder<Void> {
```

Comment:
it could be better don't throw an exception by the builder, but rather traverse the tree and return all issues (`ConstraintBuilder<Pair<Constraint, Exception>>` or something like this). It would be a concern of another place in the code what to do with this and how to represent the error.

Reply:
We assume that the user does not try to access something they are not allowed to access. This is just a security measure, not an informative validator.